lightbulb/localization.py

```diff
@@ -0,0 +1,128 @@
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+from __future__ import annotations
+
+__all__ = ["localization_unsupported", "DictLocalizationProvider", "GnuLocalizationProvider"]
+
+import collections
+import dataclasses
+import pathlib
+import typing as t
+
+import hikari
+
+from lightbulb import exceptions
+
+LocalizationMappingT: t.TypeAlias = t.Mapping[hikari.Locale, str]
+LocalizationProviderT: t.TypeAlias = t.Callable[[str], LocalizationMappingT]
+
+
+def _(string: str) -> str:
+    """
+    Utility function to fool ``xgettext`` into generating a correct ``.po`` file
+    from command and option localization keys.
+
+    Just returns the value that is passed in with no processing.
+    """
+    return string
+
+
+def localization_unsupported(_: str) -> t.NoReturn:
+    """
+    Default localization provider. Functions to disable the ability to localize commands and options. If
+    you want to use localized commands in your application you should specify a different built-in or custom
+    localization provider.
+    """
+    raise exceptions.LocalizationFailedException("no localization provider available - localization is not supported")
+
+
+@dataclasses.dataclass(slots=True, frozen=True)
+class DictLocalizationProvider:
+    """Basic localization provider that supplies localizations from a single dictionary."""
+
+    localizations: t.Mapping[hikari.Locale, t.Mapping[str, str]]
+    """Mapping containing the localizations that can be provided."""
+
+    def __call__(self, key: str) -> LocalizationMappingT:
+        out: t.Dict[hikari.Locale, str] = {}
+        for locale, translations in self.localizations.items():
+            if key in translations:
+                out[locale] = translations[key]
+
+        return out
+
+
+@dataclasses.dataclass(slots=True)
+class GnuLocalizationProvider:
+    """
+    Localization provider that parses localizations from gnu gettext compatible '.po' or '.mo' file paths. This
+    provider expects localizations to be available from the file structure expected by gettext.
+
+    I.e. ``{directory}/{locale}/{category}/{file}.[po|mo]`` (``translations/en-GB/LC_MESSAGES/commands.po``)
+    """
+
+    filename: str
+    """The name of the file containing command translations."""
+    directory: str = "translations"
+    """The base directory where the locale directories can be found."""
+    category: str = "LC_MESSAGES"
+    """The category that the translation file can be found in. Defaults to 'LC_MESSAGES'."""
+
+    _dict_provider: DictLocalizationProvider = dataclasses.field(init=False, repr=False)
+
+    def __post_init__(self) -> None:
+        try:
+            import polib
```

Comment:
Is `polib` faster than stdlib's `gettext` which works with the same file/directory structure? I don't doubt that it's probably _easier_ to work with